app/src/org/odk/collect/android/activities/FormEntryActivity.java

```diff
@@ -614,34 +522,177 @@ protected void onActivityResult(int requestCode, int resultCode, Intent intent)
         }
     }
 
-    private IntentWidget getPendingIntentWidget() {
-        IntentWidget bestMatch = null;
+    /**
+     * Performs any necessary relocating and scaling of an image coming from either a
+     * SignatureWidget or ImageWidget (capture or choose)
+     *
+     * @param originalImage the image file returned by the image capture or chooser intent
+     * @param shouldScale if false, indicates that the image is from a signature capture, so should
+     *                not attempt to scale
+     * @return the image file that should be displayed on the device screen when this question
+     * widget is in view
+     */
+    private File moveAndScaleImage(File originalImage, boolean shouldScale) throws IOException {
+        // We want to save our final image file in the instance folder for this form, so that it
+        // gets sent to HQ with the form
+        String instanceFolder =
+                mInstancePath.substring(0, mInstancePath.lastIndexOf("/") + 1);
+        String imageFilename = System.currentTimeMillis() + ".jpg";
+        String finalFilePath = instanceFolder + imageFilename;
+
+        boolean savedScaledImage = false;
+        if (shouldScale) {
+            ImageWidget currentWidget = (ImageWidget)getPendingWidget();
+            int maxDimen = currentWidget.getMaxDimen();
+            if (maxDimen != -1) {
+                savedScaledImage = FileUtils.scaleImage(originalImage, finalFilePath, maxDimen);
+            }
+        }
 
-        //Ugh, copied from the odkview mostly, that's stupid
-        for(QuestionWidget q : ((ODKView)mCurrentView).getWidgets()) {
-            //Figure out if we have a pending intent widget
-            if (q instanceof IntentWidget) {
-                if(((IntentWidget) q).getFormId().equals(mFormController.getPendingCalloutFormIndex())) {
-                    bestMatch = (IntentWidget)q;
-                }
+        File fileToReturn;
+        if (!savedScaledImage) {
+            // If we didn't create a scaled image and save it to the final path, then relocate the
+            // original image from the temp filepath to our final path
+            File finalFile = new File(finalFilePath);
+            if (!originalImage.renameTo(finalFile)) {
+                throw new IOException("Failed to rename " + originalImage.getAbsolutePath() +
+                        " to " + finalFile.getAbsolutePath());
+            } else {
+                fileToReturn = finalFile;
```

Comment:
If there's nothing blocking it, it would reduce the complexity of this code to return here and below directly, rather than letting htem both fall through to the bottom case.